http/cookies.lua

```diff
@@ -0,0 +1,353 @@
+	-- original one.
+	-- https://tools.ietf.org/html/draft-ietf-httpbis-cookie-same-site-00#section-4.1
+	if data.same_site then
+		local v
+		if data.same_site:lower() == "strict" then
+			v = "; SameSite=Strict"
+		elseif data.same_site:lower() == "lax" then
+			v = "; SameSite=Lax"
+		else
+			error('invalid value for same_site, expected "Strict" or "Lax"')
+		end
+		cookie[#cookie + 1] = v
+	end
+	return table.concat(cookie)
+end
+
+local Cookie_anchored = http_patts.Cookie * lpeg.P(-1)
+
+local function match_cookies(cookie)
+	local match = Cookie_anchored:match(cookie)
+	if match then
+		return match
+	else
+		return nil, "improper Cookie header format"
+	end
+end
+
+local function parse_cookies(cookie)
+	local cookies = match_cookies(cookie)
+	local to_add = {}
+	for k, v in pairs(cookies) do
+		to_add[#to_add + 1] = {k, v}
+	end
+	local len = #cookies
+	for i, v in ipairs(to_add) do
+		cookies[len + i] = v
+	end
+	table.sort(cookies, function(t1, t2)
+		return t1[1] < t2[1]
+	end)
+	return cookies
+end
+
+local cookiejar_methods = {}
+if psl.latest then
+	cookiejar_methods.psl_object = psl.latest()
+else
+	-- older versions of libpsl do not offer a `latest` list
+	cookiejar_methods.psl_object = psl.builtin()
+end
+local cookiejar_mt = {
+	__name = "http.cookies.cookiejar";
+	__index = cookiejar_methods;
+}
+
+local function new_cookiejar()
+	return setmetatable({cookies={}}, cookiejar_mt)
+end
+
+function cookiejar_methods:add(cookie, time)
+	cookie.last_access = time or os.time()
+	local domain, path, key = cookie.domain, cookie.path, cookie.key
+	local cookies = self.cookies
+	if cookies[domain] and cookies[domain][path] then
+		local old_cookie = cookies[domain][path][key]
+		if old_cookie then
+			cookie.creation = old_cookie.creation
+		end
+	end
+
+	local old_cookie = self:get(cookie.domain, cookie.path, cookie.key)
+	if old_cookie then
+		self:remove_cookie(old_cookie)
+	end
+	if cookie.persistent then
+		local cookie_exp_time = cookie.expires
+		local inserted = false
+		for i=1, #cookies do
+			-- insert into first spot where cookie expires after
+			if cookies[i].expires < cookie_exp_time then
+				inserted = true
+				table.insert(cookies, i, cookie)
+			end
+		end
+		if not inserted then
+			cookies[#cookies + 1] = cookie
+		end
+	else
+		cookie.expires = math.huge
+		table.insert(cookies, 1, cookie)
+	end
+
+	local by_domain = cookies[domain]
+	if not by_domain then
+		by_domain = {}
+		cookies[domain] = by_domain
+	end
+	local by_path = by_domain[path]
+	if not by_path then
+		by_path = {}
+		by_domain[path] = by_path
+	end
+	by_path[key] = cookie
+	return true
+end
+
+function cookiejar_methods:get(domain, path, key)
+	local by_domain = self.cookies[domain]
+	if not by_domain then
+		return nil
+	end
+	local by_path = by_domain[path]
+	if not by_path then
+		return nil
+	end
+	return by_path[key]
+end
+
+function cookiejar_methods:remove_cookie(cookie)
+	local cookies = self.cookies
+	for i=1, #cookies do
```

Comment:
*TODO:* Implement faster search; since cookies are sorted using time, it's possible to optimize by "leaping" through until the time is - either less than or greater than, can't recall ATM - and then go back one "leap" and going from there.